cpp/locality/NeighborList.cc

```diff
@@ -6,185 +6,201 @@
+        int last_index(-1);
+        unsigned int counter(0);
+        for (unsigned int i = 0; i < getNumBonds(); i++)
+        {
+            index = m_neighbors(i, 0);
+            if (index != last_index)
+            {
+                m_segments[index] = i;
+                if (index > 0)
+                {
+                    if (last_index >= 0)
+                    {
+                        m_counts[last_index] = counter;
+                    }
+                    counter = 0;
+                }
+            }
+            last_index = index;
+            counter++;
+        }
+        if (last_index >= 0)
+        {
+            m_counts[last_index] = counter;
+        }
+        m_segments_counts_updated = true;
+    }
 }
 
-size_t NeighborList::filter(const bool* filt)
+unsigned int NeighborList::filter(const bool* filt)
 {
     // number of good (unfiltered-out) elements so far
-    size_t num_good(0);
-    size_t* neighbors(m_neighbors.get());
-    float* weights(m_weights.get());
+    unsigned int num_good(0);
+    const unsigned int old_size(getNumBonds());
 
-    for (size_t i(0); i < m_num_bonds; ++i)
+    for (unsigned int i(0); i < getNumBonds(); ++i)
     {
         if (filt[i])
         {
-            neighbors[2 * num_good] = neighbors[2 * i];
-            neighbors[2 * num_good + 1] = neighbors[2 * i + 1];
-            weights[num_good] = weights[i];
+            m_neighbors(num_good, 0) = m_neighbors(i, 0);
+            m_neighbors(num_good, 1) = m_neighbors(i, 1);
+            m_weights[num_good] = m_weights[i];
+            m_distances[num_good] = m_distances[i];
             ++num_good;
         }
     }
-
-    const size_t old_size(m_num_bonds);
-    m_num_bonds = num_good;
-    return (size_t) num_good - old_size;
+    resize(num_good);
+    return old_size - num_good;
 }
 
-size_t NeighborList::filter_r(const freud::box::Box& box, const vec3<float>* r_i, const vec3<float>* r_j,
-                              float r_max, float r_min)
+unsigned int NeighborList::filter_r(float r_max, float r_min)
 {
-    // number of good (unfiltered-out) elements so far
-    size_t num_good(0);
-    size_t* neighbors(m_neighbors.get());
-    float* weights(m_weights.get());
-    float* distances(m_distances.get());
-
-    for (size_t bond(0); bond < m_num_bonds; ++bond)
+    // Can't use vector<bool> because that is specialized for compact storage
+    std::unique_ptr<bool[]> dist_filter(new bool[getNumBonds()]);
+    for (unsigned int i(0); i < getNumBonds(); ++i)
     {
-        const bool good(distances[bond] > r_min && distances[bond] < r_max);
-        if (good)
-        {
-            neighbors[2 * num_good] = neighbors[2 * bond];
-            neighbors[2 * num_good + 1] = neighbors[2 * bond + 1];
-            weights[num_good] = weights[bond];
-            distances[num_good] = distances[bond];
-            ++num_good;
-        }
+        dist_filter[i] = (m_distances[i] >= r_min && m_distances[i] < r_max);
```

Comment:
This is a change from existing behavior. It's very minor, but while we're looking at this is there any reason to prefer one convention over another?

Reply:
The default `r_min` is zero, and if the user doesn't provide a minimum, then I expect to retain bonds with distance 0.

Reply:
Makes sense.